extensions/auth/portability-auth-harness-microsoft/src/main/java/org/dataportabilityproject/auth/microsoft/harness/AuthTestDriver.java

```diff
@@ -0,0 +1,89 @@
+package org.dataportabilityproject.auth.microsoft.harness;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.Response;
+import okhttp3.ResponseBody;
+import org.dataportabilityproject.auth.microsoft.MicrosoftAuthDataGenerator;
+import org.dataportabilityproject.spi.gateway.types.AuthFlowConfiguration;
+import org.dataportabilityproject.types.transfer.auth.TokenAuthData;
+
+import java.awt.*;
+import java.io.IOException;
+import java.net.URI;
+import java.util.Objects;
+
+import static java.lang.System.getProperty;
+
+/**
+ *
+ */
+public class AuthTestDriver {
+    private String clientId;
+    private String secret;
+
+    private String callbackHost;
+    private String callbackBase;
+    private String authRetrievalUrl;
+
+    public AuthTestDriver() {
+        this.callbackHost = Objects.requireNonNull(getProperty("callbackHost"), "Callback host");
+        String callbackPort = Objects.requireNonNull(getProperty("callbackPort"), "Callback port");
+        this.clientId = Objects.requireNonNull(getProperty("clientId"), "Client ID");
+        this.secret = Objects.requireNonNull(getProperty("secret"), "Client secret");
+
+        callbackBase = "https://" + callbackHost + ":" + callbackPort;
+        authRetrievalUrl = callbackBase + "/code";
+    }
+
+    public static void main(String... args) throws Exception {
+        new AuthTestDriver().getOAuthTokenCode();
+    }
+
+    /**
+     * Performs an OAuth flow using the MicrosoftAuthDataGenerator, returning a token.
+     *
+     * @return the token
+     */
+    public TokenAuthData getOAuthTokenCode() throws Exception {
+
+        OkHttpClient client = TestHelper.createTestBuilder(callbackHost).build();
+        ObjectMapper mapper = new ObjectMapper();
+
+        MicrosoftAuthDataGenerator dataGenerator = new MicrosoftAuthDataGenerator("/response", () -> clientId, () -> secret, client, mapper);
+
+        AuthFlowConfiguration configuration = dataGenerator.generateConfiguration(callbackBase, "1");
+
+        Desktop desktop = Desktop.getDesktop();
+
+        desktop.browse(new URI(configuration.getUrl()));
+
+        // Execute the request and retrieve the auth code.
+        String authCode = retrieveAuthCode(client);
+
+        // get the token
+        TokenAuthData tokenData = dataGenerator.generateAuthData(callbackBase, authCode, "1", configuration.getInitialAuthData(), null);
+
+        System.out.println("TOKEN: " + tokenData.getToken());
```

Comment:
Should we bring in the logging framework here and use it.

Reply:
Yeah these are just temporary debug statements. 

One thing I would like to talk about is externalizing logging. This would involve defining an interface and having the code performing the process bootstrap (e.g. a launcher) provide an implementation. For example, the interface could be something like this:

https://gist.github.com/jimmarino/2652d5adbb3216eaed302637d0f15870

Classes that need to log data would declare the interface as a dependency and have it injected or provided by the instantiating class. 

This would allow people to easily create their own logging implementations (for example, one that uses JDK logging or a cloud logging service, or something else) without having to deal with trying to bind things to SLF4J or writing a JDK logging handler.